File: 0239_sliding_window_maximum/monotonic_deque/go/solution.go
package slidingwindowmaximum

func maxSlidingWindow(nums []int, k int) []int {
	deque := make([]int, 0, k)
	result := make([]int, 0, len(nums)-k+1)

	for i, value := range nums {
		if len(deque) > 0 && deque[0] <= i-k {
			deque = deque[1:]
		}

		for len(deque) > 0 && nums[deque[len(deque)-1]] <= value {
			deque = deque[:len(deque)-1]
		}

		deque = append(deque, i)

		if i >= k-1 {
			result = append(result, nums[deque[0]])
		}
	}

	return result
}

File: 0239_sliding_window_maximum/monotonic_deque/kotlin/solution.kt
fun maxSlidingWindow(nums: IntArray, k: Int): IntArray {
    val deque = ArrayDeque<Int>()
    val result = mutableListOf<Int>()

    for ((i, value) in nums.withIndex()) {
        while (deque.isNotEmpty() && deque.first() <= i - k) {
            deque.removeFirst()
        }

        while (deque.isNotEmpty() && nums[deque.last()] <= value) {
            deque.removeLast()
        }

        deque.addLast(i)

        if (i >= k - 1) {
            result.add(nums[deque.first()])
        }
    }

    return result.toIntArray()
}
